internal/provider/resource_access_offer.go

```diff
@@ -0,0 +1,547 @@
+			return
+		}
+	}
+
+	// Get the users to grant consume
+	var consumeUsers []string
+	if !plan.ConsumeUsers.IsNull() {
+		resp.Diagnostics.Append(plan.ConsumeUsers.ElementsAs(ctx, &consumeUsers, false)...)
+		if resp.Diagnostics.HasError() {
+			return
+		}
+	}
+
+	// Get the users to grant read
+	var readUsers []string
+	if !plan.ReadUsers.IsNull() {
+		resp.Diagnostics.Append(plan.ReadUsers.ElementsAs(ctx, &readUsers, false)...)
+		if resp.Diagnostics.HasError() {
+			return
+		}
+	}
+
+	// validate if there are overlaps or admin user
+	// validation is done here considering dynamic (juju_user resource) and static values for users
+	err := validateNoOverlapsNoAdmin(adminUsers, consumeUsers, readUsers)
+	if err != nil {
+		resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to create access offer resource, got error: %s", err))
+		return
+	}
+
+	// Call Offers.GrantOffer
+	totalUsers := make(map[permission.Access][]string)
+	totalUsers[permission.ConsumeAccess] = consumeUsers
+	totalUsers[permission.ReadAccess] = readUsers
+	totalUsers[permission.AdminAccess] = adminUsers
+
+	for access, users := range totalUsers {
+		err := a.client.Offers.GrantOffer(&juju.GrantRevokeOfferInput{
+			Users:    users,
+			Access:   string(access),
+			OfferURL: plan.OfferURL.ValueString(),
+		})
+		if err != nil {
+			resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to create access offer resource, got error: %s", err))
+			return
+		}
+	}
+
+	// Set ID as the offer URL
+	plan.ID = plan.OfferURL
+
+	// Set the plan onto the Terraform state
+	resp.Diagnostics.Append(resp.State.Set(ctx, &plan)...)
+}
+
+// Read reads users and permissions granted to the offer
+func (a *accessOfferResource) Read(ctx context.Context, req resource.ReadRequest, resp *resource.ReadResponse) {
+	// Check first if the client is configured
+	if a.client == nil {
+		addClientNotConfiguredError(&resp.Diagnostics, "access offer", "read")
+		return
+	}
+	var state accessOfferResourceOffer
+
+	// Get the Terraform state
+	resp.Diagnostics.Append(req.State.Get(ctx, &state)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Get information from ID
+	offerURL := state.ID.ValueString()
+
+	// Get user/access info from Offer
+	response, err := a.client.Offers.ReadOffer(&juju.ReadOfferInput{
+		OfferURL: offerURL,
+	})
+	if err != nil {
+		resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to read offer %s, got error: %s", offerURL, err))
+		return
+	}
+	a.trace(fmt.Sprintf("read juju offer %q", offerURL))
+
+	// Create the map
+	users := make(map[permission.Access][]string)
+	users[permission.ConsumeAccess] = []string{}
+	users[permission.ReadAccess] = []string{}
+	users[permission.AdminAccess] = []string{}
+	for _, offerUserDetail := range response.Users {
+		if offerUserDetail.UserName == "everyone@external" || offerUserDetail.UserName == "admin" {
+			continue
+		}
+
+		if _, ok := users[offerUserDetail.Access]; !ok {
```

Comment:
Have you tested the import functionality? Because the import doesn't accept a list of users/access levels, so if any users already have access the Read() which I believe is called on import, will fail.

Reply:
I didn't follow. Why would it fail?

Reply:
I think I misunderstood this line, but I was considering a scenario where the plan states "no users have access to offer foo" but when you import the offer_access, Juju is aware of a 1 user with access (e.g. Alice) then the Read() method would fail because "User Alice has unexpected access..."

But now I see this line is just checking that none of the access levels are outside of the know enumerated values. Ideally there would be a better method for this like `ValidAccessLevel()` no big deal here.